src/v/kafka/protocol/schemata/flex_api_version_generator.py

```diff
@@ -0,0 +1,174 @@
+# https://github.com/redpanda-data/redpanda/pull/4513
+#
+
+import io
+import sys
+import os
+import json
+import pathlib
+import jinja2
+
+# The only request that is never flexible is sasl_handshake_request - 17
+# Older versions of our schema may also contain values of 'none'
+never_flexible = -1
+
+# Not every value from 0 -> max_api_key is a valid request, non-supported
+# requests will map to a value of api_version(-2)
+invalid_api = -2
+
+
+class VersionRange:
+    """
+    A version range is fundamentally a range [min, max] but there are several
+    different ways in the kafka schema format to specify the bounds.
+    """
+    def __init__(self, spec):
+        self.min, self.max = self._parse(spec)
+
+    def _parse(self, spec):
+        match = re.match("^(?P<min>\d+)$", spec)
+        if match:
+            min = int(match.group("min"))
+            return min, min
+
+        match = re.match("^(?P<min>\d+)\+$", spec)
+        if match:
+            min = int(match.group("min"))
+            return min, None
+
+        match = re.match("^(?P<min>\d+)\-(?P<max>\d+)$", spec)
+        if match:
+            min = int(match.group("min"))
+            max = int(match.group("max"))
+            return min, max
+
+
+def parse_schema(schema):
+    s = io.StringIO()
+    with open(schema, 'r') as f:
+        for line in f.readlines():
+            line = re.sub("\/\/.*", "", line)
+            if line.strip():
+                s.write(line)
+
+    return json.loads(s.getvalue())
+
+
+def parse_version(f):
+    assert pathlib.Path(f).suffix == '.json'
+    schema = parse_schema(f)
+    assert schema is not None
+    api_key = schema["apiKey"]
+    assert api_key is not None
+    flex = schema['flexibleVersions']
+    assert flex is not None
+    if flex == "none":
+        flex = never_flexible
+    else:
+        r = VersionRange(flex)
+        flex = r.min
+    return api_key, flex
+
+
+def write_versions(flex_version_map, outfile):
+    """
+    For better performance generate code as a flat static array
+    """
+    assert len(flex_version_map) > 0
+    max_api_key = -1
+    for api_key, api_version in flex_version_map.items():
+        max_api_key = max(max_api_key, api_key)
+    arr_mapping = [invalid_api for _ in range(0, max_api_key + 1)]
+    for api_key, api_version in flex_version_map.items():
+        arr_mapping[api_key] = api_version
+
+    cpp_template = """#pragma once
+#include <array>
+#include "kafka/types.h"
+
+// autogenerated file - do not manually edit
+//
+// The following std::array contains a mapping between each kafka api request
+// where the api key is the index into this array which maps to the corresponding
+// first version for which flex support was introduced for that request.
+
+namespace kafka {
+class flex_versions {
+public:
+    static constexpr bool is_flexible_request(api_key key, api_version version) {
+        const api_version first_flex_version = _flex_mapping.at(key());
```

Comment:
why use array bounds checks on generated code. If any this is the only place to use fast code `map[key()]` no?

Reply:
Good point, also bounds check is done by invoking `is_in_schema` , will change +1